File: app/src/main/java/com/juniperphoton/myersplash/activity/MainActivity.java
package com.juniperphoton.myersplash.activity;

import android.animation.Animator;
import android.content.Intent;
import android.graphics.RectF;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.AppBarLayout;
import android.support.design.widget.CoordinatorLayout;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.view.ViewPager;
import android.view.View;
import android.view.ViewAnimationUtils;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.juniperphoton.myersplash.R;
import com.juniperphoton.myersplash.adapter.MainListFragmentAdapter;
import com.juniperphoton.myersplash.common.Constant;
import com.juniperphoton.myersplash.event.ScrollToTopEvent;
import com.juniperphoton.myersplash.fragment.MainListFragment;
import com.juniperphoton.myersplash.model.UnsplashCategory;
import com.juniperphoton.myersplash.model.UnsplashImage;
import com.juniperphoton.myersplash.utils.AnimatorListenerImpl;
import com.juniperphoton.myersplash.utils.DeviceUtil;
import com.juniperphoton.myersplash.utils.LocalSettingHelper;
import com.juniperphoton.myersplash.utils.RequestUtil;
import com.juniperphoton.myersplash.widget.ImageDetailView;
import com.juniperphoton.myersplash.widget.PivotTitleBar;

import org.greenrobot.eventbus.EventBus;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

import static com.juniperphoton.myersplash.utils.DisplayUtil.getDimenInPixel;

@SuppressWarnings("UnusedDeclaration")
public class MainActivity extends BaseActivity implements ImageDetailView.StateListener,
        MainListFragment.Callback {

    private static final String TAG = MainActivity.class.getName();

    private static final int SEARCH_ID = -10000;

    @BindView(R.id.pivot_title_bar)
    PivotTitleBar mPivotTitleBar;

    @SuppressWarnings("UnusedDeclaration")
    @BindView(R.id.toolbar_layout)
    AppBarLayout mAppBarLayout;

    @SuppressWarnings("UnusedDeclaration")
    @BindView(R.id.activity_main_cl)
    CoordinatorLayout mCoordinatorLayout;

    @BindView(R.id.content_activity_search_fab)
    FloatingActionButton mSearchFAB;

    @BindView(R.id.activity_main_detail_view)
    ImageDetailView mDetailView;

    @BindView(R.id.activity_main_search_view)
    com.juniperphoton.myersplash.widget.SearchView mSearchView;

    @BindView(R.id.view_pager)
    ViewPager mViewPager;

    @BindView(R.id.main_search_tag)
    TextView mTagView;

    private MainListFragmentAdapter mMainListFragmentAdapter;

    private String mQuery;
    private boolean mHandleShortcut;
    private int mDefaultIndex = 1;
    private int mLastX;
    private int mLastY;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);

        handleShortcutsAction();

        initMainViews();
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
    }

    @Override
    protected void onStart() {
        super.onStart();
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
    }

    @Override
    protected void onResume() {
        super.onResume();
        boolean scrollBar = LocalSettingHelper.getBoolean(this, Constant.SCROLL_TOOLBAR, true);
        AppBarLayout.LayoutParams params =
                (AppBarLayout.LayoutParams) mPivotTitleBar.getLayoutParams();
        if (scrollBar) {
            params.setScrollFlags(AppBarLayout.LayoutParams.SCROLL_FLAG_SCROLL
                    | AppBarLayout.LayoutParams.SCROLL_FLAG_ENTER_ALWAYS);
        } else {
            params.setScrollFlags(0);
        }

        mDetailView.registerEventBus();
        mSearchView.registerEventBus();

        mPivotTitleBar.setLayoutParams(params);

        RequestUtil.checkAndRequest(MainActivity.this);
    }

    @Override
    protected void onPause() {
        super.onPause();
        mDetailView.unregisterEventBus();
        mSearchView.unregisterEventBus();
    }

    @OnClick(R.id.content_activity_search_fab)
    void onClickSearchFAB() {
        toggleSearchView(true, true);
    }

    private void toggleSearchView(final boolean show, boolean useAnimation) {
        if (show) {
            mSearchFAB.hide();
        } else {
            mSearchFAB.show();
        }
        int[] location = new int[2];
        mSearchFAB.getLocationOnScreen(location);

        if (show) {
            mLastX = (int) (location[0] + mSearchFAB.getWidth() / 2f);
            mLastY = (int) (location[1] + mSearchFAB.getHeight() / 2f);
        }

        int width = getWindow().getDecorView().getWidth();
        int height = getWindow().getDecorView().getHeight();

        int radius = (int) (Math.sqrt(Math.pow(width, 2) + Math.pow(height, 2)));
        Animator animator = ViewAnimationUtils.createCircularReveal(mSearchView, mLastX, mLastY, show ? 0 : radius, show ? radius : 0);
        animator.addListener(new AnimatorListenerImpl() {
            @Override
            public void onAnimationEnd(Animator animation) {
                if (!show) {
                    mSearchView.reset();
                    mSearchView.setVisibility(View.GONE);
                } else {
                    mSearchView.onShown();
                }
            }
        });
        mSearchView.setVisibility(View.VISIBLE);
        if (show) {
            mSearchView.showKeyboard();
            mSearchView.onShowing();
        } else {
            mSearchView.onHiding();
        }
        if (useAnimation) {
            animator.start();
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        mDetailView.deleteShareFileInDelay();
    }

    private int getIdByIndex(int index) {
        switch (index) {
            case 0:
                return UnsplashCategory.FEATURED_CATEGORY_ID;
            case 1:
                return UnsplashCategory.NEW_CATEGORY_ID;
            case 2:
                return UnsplashCategory.RANDOM_CATEGORY_ID;
            default:
                return UnsplashCategory.NEW_CATEGORY_ID;
        }
    }

    private void initMainViews() {
        mDetailView.setNavigationCallback(this);
        mPivotTitleBar.setOnClickTitleListener(new PivotTitleBar.OnClickTitleListener() {
            @Override
            public void onSingleTap(int index) {
                if (mViewPager != null) {
                    mViewPager.setCurrentItem(index);
                    EventBus.getDefault().post(new ScrollToTopEvent(getIdByIndex(index), false));
                }
            }

            @Override
            public void onDoubleTap(int index) {
                if (mViewPager != null) {
                    mViewPager.setCurrentItem(index);
                    EventBus.getDefault().post(new ScrollToTopEvent(getIdByIndex(index), true));
                }
            }
        });
        mPivotTitleBar.setSelected(mDefaultIndex);

        mMainListFragmentAdapter = new MainListFragmentAdapter(this, getSupportFragmentManager());
        mViewPager.setAdapter(mMainListFragmentAdapter);
        mViewPager.setCurrentItem(mDefaultIndex);
        mViewPager.setOffscreenPageLimit(2);
        mViewPager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
            }

            @Override
            public void onPageSelected(int position) {
                mPivotTitleBar.setSelected(position);
                mTagView.setText("# " + mPivotTitleBar.getSelectedString());
            }

            @Override
            public void onPageScrollStateChanged(int state) {

            }
        });

        if (!DeviceUtil.hasNavigationBar(this)) {
            RelativeLayout.LayoutParams params = (RelativeLayout.LayoutParams) mSearchFAB.getLayoutParams();
            params.setMargins(0, 0, getDimenInPixel(24, this), getDimenInPixel(24, this));
            mSearchFAB.setLayoutParams(params);
        }

        mAppBarLayout.addOnOffsetChangedListener(new AppBarLayout.OnOffsetChangedListener() {
            @Override
            public void onOffsetChanged(AppBarLayout appBarLayout, int verticalOffset) {
                if ((Math.abs(verticalOffset) - appBarLayout.getHeight()) == 0) {
                    mTagView.animate().alpha(1f).setDuration(300).start();
                    mSearchFAB.hide();
                } else {
                    mTagView.animate().alpha(0f).setDuration(100).start();
                    mSearchFAB.show();
                }
            }
        });

        mTagView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                EventBus.getDefault().post(new ScrollToTopEvent(getIdByIndex(mPivotTitleBar.getSelectedItem()), false));
            }
        });
    }

    private void handleShortcutsAction() {
        if (mHandleShortcut) {
            return;
        }
        String action = getIntent().getAction();
        if (action != null) {
            switch (action) {
                case "action.search":
                    mHandleShortcut = true;
                    mAppBarLayout.post(new Runnable() {
                        @Override
                        public void run() {
                            toggleSearchView(true, false);
                        }
                    });
                    break;
                case "action.download":
                    Intent intent = new Intent(this, ManageDownloadActivity.class);
                    startActivity(intent);
                    break;
                case "action.random":
                    mHandleShortcut = true;
                    mDefaultIndex = 2;
                    break;
            }
        }
    }

    @Override
    public void onShowing() {
        mSearchFAB.hide();
    }

    @Override
    public void onHiding() {

    }

    @Override
    public void onShown() {

    }

    @Override
    public void onHidden() {
        mSearchFAB.show();
        if (mAppBarLayout.getHeight() - Math.abs(mAppBarLayout.getTop()) < 0.01) {
            mTagView.animate().alpha(1f).setDuration(300).start();
        }
    }

    @Override
    public void onBackPressed() {
        if (mSearchView.getVisibility() == View.VISIBLE) {
            if (mSearchView.tryHide()) {
                return;
            }
            toggleSearchView(false, true);
            return;
        }
        if (mDetailView.tryHide()) {
            return;
        }
        super.onBackPressed();
    }

    @Override
    public void onScrollHide() {
    }

    @Override
    public void onScrollShow() {
    }

    @Override
    public void clickPhotoItem(RectF rectF, UnsplashImage unsplashImage, View itemView) {
        int[] location = new int[2];
        mTagView.getLocationOnScreen(location);
        if (rectF.top <= (location[1] + mTagView.getHeight())) {
            mTagView.animate().alpha(0f).setDuration(100).start();
        }
        mDetailView.showDetailedImage(rectF, unsplashImage, itemView);
    }
}

File: app/src/main/java/com/juniperphoton/myersplash/activity/MainActivity.kt
package com.juniperphoton.myersplash.activity

import android.animation.Animator
import android.content.Intent
import android.graphics.RectF
import android.os.Bundle
import android.support.design.widget.AppBarLayout
import android.support.design.widget.CoordinatorLayout
import android.support.design.widget.FloatingActionButton
import android.support.v4.view.ViewPager
import android.view.View
import android.view.ViewAnimationUtils
import android.widget.RelativeLayout
import android.widget.TextView

import com.juniperphoton.myersplash.R
import com.juniperphoton.myersplash.adapter.MainListFragmentAdapter
import com.juniperphoton.myersplash.common.Constant
import com.juniperphoton.myersplash.event.ScrollToTopEvent
import com.juniperphoton.myersplash.fragment.MainListFragment
import com.juniperphoton.myersplash.model.UnsplashCategory
import com.juniperphoton.myersplash.model.UnsplashImage
import com.juniperphoton.myersplash.utils.AnimatorListenerImpl
import com.juniperphoton.myersplash.utils.DeviceUtil
import com.juniperphoton.myersplash.utils.DisplayUtil
import com.juniperphoton.myersplash.utils.LocalSettingHelper
import com.juniperphoton.myersplash.utils.PermissionUtil
import com.juniperphoton.myersplash.widget.ImageDetailView
import com.juniperphoton.myersplash.widget.PivotTitleBar

import org.greenrobot.eventbus.EventBus

import butterknife.BindView
import butterknife.ButterKnife
import butterknife.OnClick
import com.juniperphoton.myersplash.widget.SearchView

class MainActivity : BaseActivity(), ImageDetailView.StateListener, MainListFragment.Callback {

    @BindView(R.id.pivot_title_bar)
    @JvmField var mPivotTitleBar: PivotTitleBar? = null

    @BindView(R.id.toolbar_layout)
    @JvmField var mAppBarLayout: AppBarLayout? = null

    @BindView(R.id.activity_main_cl)
    @JvmField var mCoordinatorLayout: CoordinatorLayout? = null

    @BindView(R.id.content_activity_search_fab)
    @JvmField var mSearchFAB: FloatingActionButton? = null

    @BindView(R.id.activity_main_detail_view)
    @JvmField var mDetailView: ImageDetailView? = null

    @BindView(R.id.activity_main_search_view)
    @JvmField var mSearchView: SearchView? = null

    @BindView(R.id.view_pager)
    @JvmField var mViewPager: ViewPager? = null

    @BindView(R.id.main_search_tag)
    @JvmField var mTagView: TextView? = null

    private var mMainListFragmentAdapter: MainListFragmentAdapter? = null

    private var mHandleShortcut: Boolean = false
    private var mDefaultIndex = 1
    private var mLastX: Int = 0
    private var mLastY: Int = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        ButterKnife.bind(this)

        handleShortcutsAction()

        initMainViews()
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
    }

    override fun onStart() {
        super.onStart()
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
    }

    override fun onResume() {
        super.onResume()
        mDetailView!!.registerEventBus()
        mSearchView!!.registerEventBus()

        PermissionUtil.checkAndRequest(this@MainActivity)
    }

    override fun onPause() {
        super.onPause()
        mDetailView!!.unregisterEventBus()
        mSearchView!!.unregisterEventBus()
    }

    @OnClick(R.id.content_activity_search_fab)
    internal fun onClickSearchFAB() {
        toggleSearchView(true, true)
    }

    private fun toggleSearchView(show: Boolean, useAnimation: Boolean) {
        if (show) {
            mSearchFAB!!.hide()
        } else {
            mSearchFAB!!.show()
        }
        val location = IntArray(2)
        mSearchFAB!!.getLocationOnScreen(location)

        if (show) {
            mLastX = (location[0] + mSearchFAB!!.width / 2f).toInt()
            mLastY = (location[1] + mSearchFAB!!.height / 2f).toInt()
        }

        val width = window.decorView.width
        val height = window.decorView.height

        val radius = Math.sqrt(Math.pow(width.toDouble(), 2.0) + Math.pow(height.toDouble(), 2.0)).toInt()
        val animator = ViewAnimationUtils.createCircularReveal(mSearchView, mLastX, mLastY, (if (show) 0 else radius).toFloat(), (if (show) radius else 0).toFloat())
        animator.addListener(object : AnimatorListenerImpl() {
            override fun onAnimationEnd(animation: Animator) {
                if (!show) {
                    mSearchView!!.reset()
                    mSearchView!!.visibility = View.GONE
                } else {
                    mSearchView!!.onShown()
                }
            }
        })
        mSearchView!!.visibility = View.VISIBLE
        if (show) {
            mSearchView!!.tryShowKeyboard()
            mSearchView!!.onShowing()
        } else {
            mSearchView!!.onHiding()
        }
        if (useAnimation) {
            animator.start()
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        mDetailView!!.deleteShareFileInDelay()
    }

    private fun getIdByIndex(index: Int): Int {
        when (index) {
            0 -> return UnsplashCategory.FEATURED_CATEGORY_ID
            1 -> return UnsplashCategory.NEW_CATEGORY_ID
            2 -> return UnsplashCategory.RANDOM_CATEGORY_ID
            else -> return UnsplashCategory.NEW_CATEGORY_ID
        }
    }

    private fun initMainViews() {
        mDetailView!!.setNavigationCallback(this)
        mPivotTitleBar!!.setOnClickTitleListener(object : PivotTitleBar.OnClickTitleListener {
            override fun onSingleTap(index: Int) {
                if (mViewPager != null) {
                    mViewPager!!.currentItem = index
                    EventBus.getDefault().post(ScrollToTopEvent(getIdByIndex(index), false))
                }
            }

            override fun onDoubleTap(index: Int) {
                if (mViewPager != null) {
                    mViewPager!!.currentItem = index
                    EventBus.getDefault().post(ScrollToTopEvent(getIdByIndex(index), true))
                }
            }
        })
        mPivotTitleBar!!.selectedItem = mDefaultIndex

        mMainListFragmentAdapter = MainListFragmentAdapter(this, supportFragmentManager)
        mViewPager!!.adapter = mMainListFragmentAdapter
        mViewPager!!.currentItem = mDefaultIndex
        mViewPager!!.offscreenPageLimit = 2
        mViewPager!!.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {
            override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {}

            override fun onPageSelected(position: Int) {
                mPivotTitleBar!!.selectedItem = position
                mTagView!!.text = "# ${mPivotTitleBar!!.selectedString}"
            }

            override fun onPageScrollStateChanged(state: Int) {

            }
        })

        if (!DeviceUtil.hasNavigationBar(this)) {
            val params = mSearchFAB!!.layoutParams as RelativeLayout.LayoutParams
            params.setMargins(0, 0,
                    DisplayUtil.getDimenInPixel(24, this),
                    DisplayUtil.getDimenInPixel(24, this))
            mSearchFAB!!.layoutParams = params
        }

        mAppBarLayout!!.addOnOffsetChangedListener { appBarLayout, verticalOffset ->
            if (Math.abs(verticalOffset) - appBarLayout.height == 0) {
                mTagView!!.animate().alpha(1f).setDuration(300).start()
                mSearchFAB!!.hide()
            } else {
                mTagView!!.animate().alpha(0f).setDuration(100).start()
                mSearchFAB!!.show()
            }
        }

        mTagView!!.setOnClickListener { EventBus.getDefault().post(ScrollToTopEvent(getIdByIndex(mPivotTitleBar!!.selectedItem), false)) }
    }

    private fun handleShortcutsAction() {
        if (mHandleShortcut) {
            return
        }
        val action = intent.action
        if (action != null) {
            when (action) {
                "action.search" -> {
                    mHandleShortcut = true
                    mAppBarLayout!!.post { toggleSearchView(true, false) }
                }
                "action.download" -> {
                    val intent = Intent(this, ManageDownloadActivity::class.java)
                    startActivity(intent)
                }
                "action.random" -> {
                    mHandleShortcut = true
                    mDefaultIndex = 2
                }
            }
        }
    }

    override fun onShowing() {
        mSearchFAB!!.hide()
    }

    override fun onHiding() {

    }

    override fun onShown() {

    }

    override fun onHidden() {
        mSearchFAB!!.show()
        if (mAppBarLayout!!.height - Math.abs(mAppBarLayout!!.top) < 0.01) {
            mTagView!!.animate().alpha(1f).setDuration(300).start()
        }
    }

    override fun onBackPressed() {
        if (mSearchView!!.visibility == View.VISIBLE) {
            if (mSearchView!!.tryHide()) {
                return
            }
            toggleSearchView(false, true)
            return
        }
        if (mDetailView!!.tryHide()) {
            return
        }
        super.onBackPressed()
    }

    override fun onScrollHide() {}

    override fun onScrollShow() {}

    override fun clickPhotoItem(rectF: RectF, unsplashImage: UnsplashImage, itemView: View) {
        val location = IntArray(2)
        mTagView!!.getLocationOnScreen(location)
        if (rectF.top <= location[1] + mTagView!!.height) {
            mTagView!!.animate().alpha(0f).setDuration(100).start()
        }
        mDetailView!!.showDetailedImage(rectF, unsplashImage, itemView)
    }

    companion object {
        private val TAG = "MainActivity"

        private val SEARCH_ID = -10000
    }
}
